From 34300339117913f7f65a2ef508f3e80dd3041f0a Mon Sep 17 00:00:00 2001
From: Keerthana K <keerthanak@vmware.com>
Date: Thu, 23 Sep 2021 13:50:58 +0000
Subject: [PATCH 5/9] stalld: Fix gcc options in Makefile

- Fix Makefile to consider spec option only if file exists.
  CFLAGS option is added to match the rpmbuild in RHEL.
  since Photon does not provide those files, fixing Makefile
  to skip this option if those files are not present.
  Enable -fstack-clash-protection and -fstack-protector
  only if gcc version is greater than 8.0.0

- Add machine dependant options based on architecture check.
  This option (-m64) sets int to 32 bits and long and pointer
  types to 64 bits, and generates code for the x86-64 architecture.
  We have an equivalent option for arm64 as well, which is -mabi=lp64.

- Add -fcf-protection option only for x86-64. This option is
  used to enable code instrumentation of control-flow transfers
  to increase program security by checking that target addresses
  of control-flow transfer instructions (such as indirect function
  call, function return, indirect jump) are valid. This prevents
  diverting the flow of control to an unexpected target. This is
  intended to protect against such threats as Return-oriented
  Programming (ROP), and similarly call/jmp-oriented programming (COP/JOP).
  This option is not supported in gcc aarch64.

Signed-off-by: Keerthana K <keerthanak@vmware.com>
---
 Makefile | 26 ++++++++++++++++++++++----
 1 file changed, 22 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index f866a74..8b53e12 100644
--- a/Makefile
+++ b/Makefile
@@ -42,11 +42,29 @@ $(info TMOPTS=$(TMOPTS))
 INSTALL	=	install
 CC	:=	gcc
 FOPTS	:=	-flto=auto -ffat-lto-objects -fexceptions -fstack-protector-strong \
-		-fasynchronous-unwind-tables -fstack-clash-protection $(strip $(FCF_PROTECTION))
-MOPTS   :=  $(strip $(TMOPTS))
+		-fasynchronous-unwind-tables
+ARCH	:=	$(shell uname -m)
+GCCVERSION:=	$(shell gcc -dumpversion | awk -F. '{print $$3+100*($$2+100*$$1)}')
+VER	:=	80000
+ifeq ($(shell expr $(GCCVERSION) \>= $(VER)), 1)
+FOPTS	+=	-fstack-clash-protection
+    ifeq ($(ARCH), x86_64)
+        FOPTS	+=	-fcf-protection
+    endif
+endif
+ifeq ($(ARCH), x86_64)
+        MOPTS	:=	-m64 -mtune=generic
+else
+    ifeq ($(ARCH), aarch64)
+        MOPTS	:=	-mabi=lp64 -mtune=generic
+    endif
+endif
 WOPTS	:= 	-Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS
-SOPTS	:= 	-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1
-
+ifneq (,$(wildcard /usr/lib/rpm/redhat/redhat-hardened-cc1))
+ifneq (,$(wildcard /usr/lib/rpm/redhat/redhat-annobin-cc1))
+    SOPTS	:=	-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1
+endif
+endif
 CFLAGS	:=	-O2 -g -DVERSION=\"$(VERSION)\" $(FOPTS) $(MOPTS) $(WOPTS) $(SOPTS) -DUSE_BPF=$(USE_BPF)
 LDFLAGS	:=	-ggdb
 
-- 
2.39.0

